connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -117,7 +131,36 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
+    responseObserver.onCompleted()
+  }
+
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val schema = dataframe.schema
+    val maxRecordsPerBatch = dataframe.sparkSession.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = dataframe.sparkSession.sessionState.conf.sessionLocalTimeZone
+
+    val batches = dataframe.queryExecution.executedPlan
+      .execute()
+      .mapPartitionsInternal { iter =>
+        ArrowConverters
+          .toArrowBatchIterator(iter, schema, maxRecordsPerBatch, timeZoneId, TaskContext.get)
+      }
+
+    batches.toLocalIterator.foreach { case (bytes, count, size) =>
```

Comment:
Let's also avoid using local iterator. That will make things unnecessarily slow because it limits parallelism. I was thinking doing the following:
```scala
def collectArrow(callback: Array[Array[Byte]] => ()): Unit = {
    withAction("collectArrow", queryExecution) { plan =>
      val rdd = toArrowBatchRdd(plan)
      val signal = new Object
      val numPartitions = rdd.getNumPartitions
      val availablePartitions = mutable.Map.empty[Int, Array[Array[Byte]]]
      def onNewPartition(partitionId: Int, batches: Array[Array[Byte]]): Unit = {
        signal.synchronized {
          availablePartitions(partitionId) = batches
          signal.notify()
        }
      }
      sparkSession.sparkContext.submitJob(
        rdd = rdd,
        processPartition = (i: Iterator[Array[Byte]]) => i.toArray,
        partitions = 0 until numPartitions,
        resultHandler = onNewPartition,
        resultFunc = () => ())

      var currentPartitionId = 0
      while (currentPartitionId < numPartitions) {
        val batches = signal.synchronized {
          var result = availablePartitions.remove(currentPartitionId)
          while (result.isEmpty) {
            signal.wait()
            result = availablePartitions.remove(currentPartitionId)
          }
          result.get
        }
        callback(batches)
        currentPartitionId += 1
      }
    }
  }
```
This was hacked together in Dataset, but IMO it is better to do this somewhere in here.

Reply:
Yeah we should avoid `toLocalIterator`

Reply:
maybe we should also add a field `partitionId` in proto message, than we sort by it in client to keep the order